log: change fingerprint_id type to string

kyle-a-wong opened the pull request:
Previously, transaction_fingerprint_id and statement_fingerprint_id were typed as uint64. Somewhere downstream, these uint64 fields were truncated into int64 max ints. As a result, telemetry logs had the same fingerprint id for different sql statements.

To fix, the field name was re-typed as a string to ensure no truncation / type conversion happens.

Epic: none
Fixes: #123665
Release note (bug fix): Fixed a bug where telemetry logs were had the same statement fingerprint id for different sql statements.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124628)
<!-- Reviewable:end -->

dhartunian reviewed (commented):
Can you add a test that encodes a long fingerprint and verify that it's not truncated? Good opportunity to use the `datadriven` library to validate the entire JSON payload.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124628#-:-NybYxEfB95IdotdSXUX:bb0hnyo)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @xinhaoz)



<!-- Sent from Reviewable.io -->

kyle-a-wong committed:
log: fix FormatEntryAsJSON in structured log spy

Previously, this `FormatEntryAsJSON` was using unmarshal
from the json package to unmarshal a json string  into a
`map[string]interface{}`. This resulted into `int64` and
`uint64` typed fields to be rounded due to how this
function handles unmarshaling these types into a
`interface{}` type.

To fix, json `Decoder` with `UseNumber` is now being used
which fixes this issue.

Release note: None

kyle-a-wong committed:
log: change fingerprint_id type to string

Previously, transaction_fingerprint_id and
statement_fingerprint_id were typed as uint64.
Somewhere downstream, these uint64 fields were
truncated into int64 max ints. As a result,
telemetry logs had the same fingerprint id for
different sql statements.

To fix, the field name was re-typed as a string
to ensure no truncation / type conversion happens.

Epic: none
Fixes: #123665
Release note (bug fix): Fixed a bug where telemetry
logs were had the same statement fingerprint id for
different sql statements.

kyle-a-wong reviewed (commented):
done

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124628#-:-Nz58c40Db9ezJO4g4iY:b-7rl0kd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @xinhaoz)



<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just a question on the commit msg motivation

do you know what will happen if we use the new code to decode a payload generated with the old code? will it just drop the fingerprint IDs or cause an error?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124628#-:-Nz9Qq2Z4T8-uIt4A-W-:b-p45yhx)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @kyle-a-wong and @xinhaoz)

___
*[`-- commits` line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124628#-Nz9R97_AHKTwfxVDWgv):*
There's a natural question that someone would ask after reading this: why not just eliminate the `int64` conversion? I think we have additional motivation in that the numbers get truncated when parsed by **other** JSON parsers such as Datadog's, right?


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
As discussed in person: the main potential concern is when debug zip is reading the logs and unmarshalling/deserializing them back into the protobuf type. Debug zip deserializes logs into the `logpb.Entry` type and log message is deserialized into `Entry.message`, which is a string type. Since the message itself isn't deserialized into SampledQuery and SampledTransaction, we shouldn't run into compatibility issues.

If we attempt to unmarshal an older payload into the updated structs, the field will just be left blank since the field types don't match, but it shouldn't cause an error

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124628#-:-NzA1BKE6A1znFk1tFuk:b-20sm1z)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dhartunian and @xinhaoz)

___
*[`-- commits` line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124628#-Nz9R97_AHKTwfxVDWgv):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

There's a natural question that someone would ask after reading this: why not just eliminate the `int64` conversion? I think we have additional motivation in that the numbers get truncated when parsed by **other** JSON parsers such as Datadog's, right?
</blockquote></details>

The main issue here is the  unmarshalling logic we used and specifically how golang handles `json.Unmarshal()` into an untyped (interface{}) maps.

 I would hope that other json parsers like Datadog's would be able to elegantly handle int64s since there are times when people may need to use int64 / longs in a json payload.


<!-- Sent from Reviewable.io -->

kyle-a-wong commented:
bors r+

kyle-a-wong commented:
bors r+
